### PR TITLE
live: `html` report without DVC.

### LINKED ISSUE #213: html generation: make dvclive independent of dvc
In iterative/dvc#6944 we are considering extracting the rendering outside of DVC.
This issue would go nicely with #163 - because we could let both DVC and dvclive use the same rendering backend, which would make dvclive more functional. 

This unravels some problems:
1. Plot is not only some data, but also a configuration
2. DVC has a capability to store the configuration, dvclive does not

In order to provide such functionality, we need to let `dvclive` provide configuration to the potential plots library.
We discussed this issue with @dberenbaum and @daavoo and it seems that we could go two ways about it:

1. Let the `dvclive` render the data only at the runtime - we know what method has been called and can provide proper config from within the library - this however would make visualizing previously logged data hard. We could avoid that by saving the config  on DVC side (similar to what we do with `html` and `summary` now) - though that would make it work only in DVC + `dvclive` use case. Standalone `dvclive` would be limited.

2. Let the `dvclive` store the configuration - that seems like our desired behavior, though it poses a problem: how to structure the plot config and how to store it so that both standalone `dvclive` and `dvvclive` + DVC use cases work properly?
Plot config would probably be defined by iterative/dvc#7086
In the same issue we are intending to define plots specification inside `dvc.yaml` - but if we want `dvclive` to render too, we would need to store it somewhere non-DVC related. Initial idea would be to store it in `dvclive_path/.config`, but in this case, when we are using `dvclive` alongside DVC we will need to make sure DVC reads additional `plots` definitions from within `dvclive` dirs.

Maybe we could let `plots` section inside `dvc.yaml` address other file by path?
eg:
```
# dvc.yaml
plots:
   plots_configs: ['dvclive/.config.yaml']
   plot1:
   ...
   plot2:
   ...
   
```

Which would let DVC know where to look for additional plots configurations and "expand it" on `dvc plots show/diff`. That would also make `dvclive` independent of DVC in terms of plots generation.
An additional problem for this use case is templates: do we want to let users define their own plots templates? 
We will have to store them somewhere.

Did I miss something? cc @dberenbaum @daavoo

### PR DESCRIPTION
**Pull request by daavoo**
- Uses `dvc_render`.
- Add new method `Live.make_report` that users can call from a `Live` instance at any point. 

  This method is still called by default on `next_step` if the report has been enabled. I'm not sure about preserving that behavior and would move towards explicit `make_report` calls.
  
- Add new entry point `dvclive-show` that can generate the latest HTML report.

Defaults to the previous behavior, for now. `html` report only gets generated if enabled on the `dvc.yaml` side . However, users can manually enable `html` without DVC by passing:

```python
live = Live(report="html")
```

Closes #213 #195

### CODE REVIEW

File: dvclive/data/plot.py
Comment (daavoo) [inline dvclive/data/plot.py:65]: All this logic is to somehow allow preserving behavior with and without DVC by serializing templates. So users can set in DVC.yaml:

```
stages:
  test-dvclive:
    cmd: python test_dvclive.py
    plots:
      - dvclive/plots/confusion_matrix.json:
          template: dvclive/plots/confusion_matrix.template
```

DVCLive would be able to work without the serialized template.

Nor sure if the templates should be dumped to a separate subfolder (i.e. `dvclive/templates` instead of `dvclive/plots`) in order to be preserved across runs:

```
stages:
  test-dvclive:
    cmd: python test_dvclive.py
    plots:
      - dvclive/plots/confusion_matrix.json:
          template: dvclive/templates/confusion_matrix.template
```

File: dvclive/live.py
Comment (daavoo) [inline dvclive/live.py:108]: This is what closes #195 . Report is now generated in `dvclive/report.html`.

File: dvclive/live.py
Comment (daavoo) [inline dvclive/live.py:186]: Users can manually call this at any point from a `Live` instance

File: dvclive/report.py
Comment (daavoo) [inline dvclive/report.py:69]: Added `CLI` utility that can reproduce the latest `html` report.

File: dvclive/data/plot.py
Comment (dberenbaum) [reply dvclive/data/plot.py:65]: Is this only necessary for `dvc plots`? Is serialization needed for dvclive to make the report?

File: dvclive/live.py
Comment (dberenbaum) [reply dvclive/live.py:30]: Should we assert and raise an error if `report` is something other than `html`?

File: dvclive/live.py
Comment (dberenbaum) [reply dvclive/live.py:186]: Hm, it seems odd that `report` is kwarg in `Live()`, but it doesn't get passed in `live.make_report()` calls (unlike `live.next_step()`). Seems like it should either default to whatever is passed in `Live()`, or `report` kwarg should be removed from `Live()`.

Related: It seems odd to pass `Live(report="html")` in no-step scenarios and have that kwarg do nothing. So maybe that's not the best place for it? Not sure 🤔 .

File: dvclive/report.py
Comment (dberenbaum) [reply dvclive/report.py:69]: This is very cool and we may need it at some point, but I think it might be overkill in this PR. Does it add enough value yet over documenting `report.html_report()` and keeping everything in the Python API?

File: dvclive/report.py
Comment (dberenbaum) [inline dvclive/report.py:59]: Not sure where to address this, but one thing that's missing from the dvc implementation is an output link and/or auto open of the report.

**Review by dberenbaum — COMMENTED** [review]
🔥 This is great @daavoo. Still some big open questions, and I want to see how it works in DVC once that is updated, but exciting to see reports working without dvc.

File: dvclive/live.py
Comment (daavoo) [reply dvclive/live.py:30]: > Should html be the default, at least for now? It matches current behavior when run with dvc.

It already matches the behavior when run with DVC, (at least while the `live` section is not deprecated) with:

https://github.com/iterative/dvclive/blob/90c46ea602a2bd04052b334fab57e6ba04940c36/dvclive/live.py#L84-L85

File: dvclive/data/plot.py
Comment (daavoo) [reply dvclive/data/plot.py:65]: > Is this only necessary for dvc plots?

Yes

> Is serialization needed for dvclive to make the report?

It's not needed, DVCLive could generate the report without dumping templates.

File: dvclive/report.py
Comment (daavoo) [reply dvclive/report.py:69]: > Does it add enough value yet over documenting report.html_report() and keeping everything in the Python API?

Probably it doesn't.

File: dvclive/live.py
Comment (daavoo) [reply dvclive/live.py:186]: I'm not sure either.

It might be better to remove `mode` from `make_report` and make the `report` arg of `Live` the only source of truth.

One problem with the alternative (removing `report` from `Live`) is that integrations become a little more complicated because right now users can customize `Live` args but in this case we would need to handle `report` separately

File: dvclive/data/plot.py
Comment (dberenbaum) [reply dvclive/data/plot.py:65]: IMO it might be better to at least put this in a follow-up PR because it might need more discussion or a different approach might come up. Even without it, we are achieving parity with current DVC behavior, right?

File: dvclive/live.py
Comment (dberenbaum) [reply dvclive/live.py:186]: > One problem with the alternative (removing `report` from `Live`) is that integrations become a little more complicated because right now users can customize `Live` args but in this case we would need to handle `report` separately

Got it, makes sense.

> It might be better to remove `mode` from `make_report` and make the `report` arg of `Live` the only source of truth.

Since it's a method of `Live`, I would expect it to use `self._report` unless I override it, so what about something like this?

```
    def make_report(self, mode: Optional[str] = None):
           if not mode:
               mode = self._report
```

Also, maybe we should rename in one place or the other so it is consistently either `report` or `mode`?

File: dvclive/report.py
Comment (dberenbaum) [inline dvclive/report.py:34]: So this changes current dvc behavior, right? @pared Do you remember why we decided to include refs to all image files instead of encoding them directly into the HTML?

File: dvclive/live.py
Comment (dberenbaum) [reply dvclive/live.py:108]: Will it be an issue if the dvclive dir is marked as a plots output and there is already an html file inside it?

Comment (daavoo) [reply dvclive/live.py:108]: > Will it be an issue if the dvclive dir is marked as a plots output and there is already an html file inside it?

Not sure if an issue but the HTML will be tracked by whichever config they set for the `plots` output (so HTML might be added to DVC cache).

It's not strictly related with this P.R. but one of the directions I also wanted with deprecating `live` was to encourage more granular output config for DVCLive outputs which is not currently possible:

```yaml
...
plots:
  - dvclive/scalars
      cache: false
  - dvclive/images
metrics:
  # I mean to open issue to discuss also nesting summary
  - dvclive/summary.json:
      cache: false
```

```
#.gitignore
dvclive/report.html
```

Comment (dberenbaum) [reply dvclive/live.py:108]: Also, looks like the dvc output will be wrong once this is merged:

```
Live summary will be created at:
/Users/dave/Code/dvc-checkpoints-mnist/dvclive_dvc_plots/index.html
```

How do we sequence these PRs to keep everything working as expected? Would it be better to close #295 in a follow-up PR?

File: dvclive/live.py
Comment (dberenbaum) [reply dvclive/live.py:30]: > It already matches the behavior when run with DVC, (at least while the `live` section is not deprecated) with:
> 
> [](https://github.com/iterative/dvclive/blob/90c46ea602a2bd04052b334fab57e6ba04940c36/dvclive/live.py#L84-L85)

True. Does it cause an issue to default to `report = "html"`? Any other reason you wouldn't want it to be the default?

AFAIK the only reason we didn't always generate it before was because we couldn't without dvc. We used to have `html = True` by default.

File: dvclive/live.py
Comment (daavoo) [reply dvclive/live.py:108]: A subsequent DVC P.R. should be created to fix that message but more importantly to deactivate the HTML generation on DVC side. 

`dvclive` is pinned in DVC so I see the sequence as:

Merging https://github.com/iterative/dvc/pull/7409 > Approval of this P.R. > New P.R. in DVC fixing message and disabling generation.

Comment (dberenbaum) [reply dvclive/live.py:108]: One other discrepancy I noticed:

https://github.com/iterative/dvc/blob/6a6955e27cf69490c7f694980dadfd3f7a5f9949/dvc/repo/live.py#L22

The dvc-generated html includes metrics, but this PR does not. Is there a plan to include metrics in a future PR?

Comment (dberenbaum) [reply dvclive/live.py:108]: > deactivate the HTML generation on DVC side

Should html still be generated now in `dvclive_dvc_plots`? I'm not seeing that happen.

Comment (dberenbaum) [reply dvclive/live.py:108]: I'm confused what's happening now in `dvclive_dvc_plots` with these changes. It doesn't look like it's being updated live anymore after this PR.

File: dvclive/report.py
Comment (pared) [reply dvclive/report.py:34]: There wasn't any necessity that would make us use ``. We decided to go this way to avoid producing gigantic htmls which would be troubling to parse/inspect.

File: dvclive/live.py
Comment (daavoo) [reply dvclive/live.py:30]: Made `html` default for all cases

File: dvclive/report.py
Comment (daavoo) [reply dvclive/report.py]: Removed

File: dvclive/report.py
Comment (daavoo) [reply dvclive/report.py:58]: Added message to `Live` init. Need to add auto open

File: dvclive/live.py
Comment (daavoo) [reply dvclive/live.py:111]: > I'm confused what's happening now in dvclive_dvc_plots with these changes. It doesn't look like it's being updated live anymore after this PR.

If you install this P.R. and https://github.com/iterative/dvc/pull/7514 (follow up for DVC once this P.R. gets merged), then :

- `dvclive_dvc_plots` should not be generated at all
- The new HTML report (with correct refreshing) should be generated at `{dvclive_dir}/report.html`.

File: dvclive/data/plot.py
Comment (daavoo) [reply dvclive/data/plot.py]: > IMO it might be better to at least put this in a follow-up PR because it might need more discussion or a different approach might come up. Even without it, we are achieving parity with current DVC behavior, right?

Removed from this P.R.

File: dvclive/live.py
Comment (pared) [reply dvclive/live.py:114]: If we are separating this, we might as well remove `dvc/commands/live` functionality. We created for visualizing `dvclive` results, but  now `dvclive` will be able to do it by itself.

File: dvclive/live.py
Comment (pared) [inline dvclive/live.py:194]: We should probably do it only in case of existing `self._report`. For example:
```
python -c "from dvclive import Live;logger=Live('ble',report=None);logger.log('value', 10);logger.next_step()"
```
will raise some errors, which I wouldn't expect, since I specified that I don't want  a report.

File: setup.py
Comment (dberenbaum) [inline setup.py:39]: Should this be optional? It seems like this is the basic functionality of dvclive.

File: setup.py
Comment (pared) [inline setup.py:85]: So I cannot say that `dvclive` has no dependencies anymore? :(

File: dvclive/report.py
Comment (pared) [inline dvclive/report.py:68]: `dvclive` gets polluted with the internal `dvc` data structure, I will create an issue on the dvc-render side to get it fixed.

**Review by pared — CHANGES_REQUESTED** [review]
One thing to fix.

File: dvclive/report.py
Comment (pared) [reply dvclive/report.py:68]: Well, i won't, its already there:
https://github.com/iterative/dvc-render/issues/19

File: setup.py
Comment (dberenbaum) [reply setup.py:85]: As long as we can say it's "lightweight" and people are able to quickly install it on any system, I don't think we should worry about it.

File: dvclive/report.py
Comment (dberenbaum) [reply dvclive/report.py:34]: Looking at this again, it seems like it could result in giant htmls if there's enough image data. It also duplicates what's already saved in `dvclive/images`. Can we refer to those image files instead? Is there a reason to prioritize encoding them?

Comment (daavoo) [reply dvclive/report.py:34]: > Can we refer to those image files instead? 

Yes, we can.

> Is there a reason to prioritize encoding them?

No strong reason, just making the HTML standalone, similar to how the plot data is embedded. If we refer to the image folder the report might break if someone remove/alter images from the folder.

TBH I think the vega embedding takes proportionally much more space per item than the base64 encoded string for each image.

File: dvclive/live.py
Comment (dberenbaum) [reply dvclive/live.py]: > It seems odd to pass Live(report="html") in no-step scenarios and have that kwarg do nothing. 

Not sure if it should be part of this PR, but currently, I can use the `live` section in `dvc.yaml` to get a report even in no-step scenarios, but it doesn't seem like it's possible without the `live` section. Is there some way to still generate the report by default?

File: dvclive/report.py
Comment (daavoo) [reply dvclive/report.py:34]: To clarify, the duplication is not introduced in this P.R. 

In the current DVC HTML, image data is being duplicated and saved in a separate directory (https://github.com/iterative/dvc/issues/6665)

File: dvclive/live.py
Comment (daavoo) [reply dvclive/live.py]: >  Is there some way to still generate the report by default?

When using the `live` section in no-step scenario, the report gets generated by the monitor task after the stage has ended. 
Any implementation on DVCLive side would behave differently, as the `Live` object doesn't have info of when the "stage" has ended (the closest would be to implement the logic in `__del__`) .
IMO would be better to document `Live.make_report()` and emphasize it's usage for non step scenarios

File: setup.py
Comment (pared) [reply setup.py:85]: Yeah, I'm just joking. Though, to be honest, there was something magical to the "It has NO dependencies". :)

**Review by pared — APPROVED** [review]

File: dvclive/report.py
Comment (dberenbaum) [reply dvclive/report.py:34]: True, although in this case we already have everything consolidated into a folder. Anyway, we can keep as is for now. No need to block the PR.

File: dvclive/live.py
Comment (dberenbaum) [reply dvclive/live.py]: Extracted to #231

**Review by dberenbaum — APPROVED** [review]

**Pull request merged**
